bin/node/e2e-tests/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+            >,
+            Self::SelectChain,
+            Self::BlockImport,
+        ),
+        sc_service::Error,
+    > {
+        let (client, backend, keystore, task_manager) =
+            new_full_parts::<Self::Block, Self::RuntimeApi, Self::Executor>(config)?;
+        let client = Arc::new(client);
+
+        let inherent_providers = InherentDataProviders::new();
+        let select_chain = sc_consensus::LongestChain::new(backend.clone());
+
+        let (grandpa_block_import, ..) =
+            grandpa::block_import(client.clone(), &(client.clone() as Arc<_>), select_chain.clone())?;
+
+        let (block_import, babe_link) = sc_consensus_babe::block_import(
+            sc_consensus_babe::Config::get_or_compute(&*client)?,
+            grandpa_block_import,
+            client.clone(),
+        )?;
+
+        let consensus_data_provider = BabeConsensusDataProvider::new(
+            client.clone(),
+            keystore.sync_keystore(),
+            &inherent_providers,
+            babe_link.epoch_changes().clone(),
+            vec![(AuthorityId::from(Alice.public()), 1000)],
+        )
+            .expect("failed to create ConsensusDataProvider");
+
+        Ok((
+            client,
+            backend,
+            keystore.sync_keystore(),
+            task_manager,
+            inherent_providers,
+            Some(Box::new(consensus_data_provider)),
+            select_chain,
+            block_import,
+        ))
+    }
+}
+
+/// performs a runtime upgrade given wasm blob and a handle to a node.
+pub fn perform_runtime_upgrade(node: &Node<NodeTemplateChainInfo>) {
+    type SystemCall = frame_system::Call<Runtime>;
+    type SudoCall = pallet_sudo::Call<Runtime>;
+
+    let whales = vec![
+        "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY", //Alice
+        "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty", //Bob
+    ]
+        .into_iter()
+        .map(|account| AccountId32::from_str(account).unwrap())
+        .collect::<Vec<_>>();
+
+    let wasm = include_bytes!("./runtime.wasm");
+
+    let call = Call::Sudo(SudoCall::sudo(Box::new(Call::System(SystemCall::set_code(wasm.to_vec())))));
+    node.submit_extrinsic(call, whales[0].clone());
+    node.seal_blocks(1);
+
+    // assert that the runtime is upgraded by looking at events
+    let events = node.with_state(|| frame_system::Module::<Runtime>::events())
+        .into_iter()
+        .filter(|event| {
+            match event.event {
+                Event::frame_system(frame_system::RawEvent::CodeUpdated) => true,
+                _ => false,
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // make sure event is in state
+    assert_eq!(events.len(), 1);
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use pallet_balances_e2e_tests::*;
+
+    #[test]
+    fn runtime_upgrade() {
+        let node = Node::<NodeTemplateChainInfo>::new().unwrap();
+
+        // first perform runtime upgrade
+        perform_runtime_upgrade(&node);
```

Comment:
if I read this correctly, this test overwrites the runtime with the same runtime with no previous state, right? So, it doesn't really test whether stuff worked before and after the upgrade, does it? Nor whether the migration works (as data structures didn't change).

How is that intended to be included?

Reply:
>  this test overwrites the runtime with the same runtime with no previous state, right? 

basically, this is just a demo/placeholder until we figure out how to set up an existing chain on CI and run actual runtime upgrades.